leapdao/leap-node
# fix: replay submissions and restore period proposals on restart

This PR solves two problems: 

## 1. When replaying the chain, period submission are not saved to local db #395 

The goal here is the same: for every Submission event happened on root chain, we want a record in a local level db database for the node keeping the event data. We index these events by period start height, so that we can later query them via API for proof construction (https://github.com/leapdao/leap-node/pull/319).

## Changes overview:
- the node were batching all the Submissions in a transient variable (`bridgeState.submissions`) and saved those every block. Now it still collects submissions, but saves each submission individually once it is found onchain (during `submitPeriod` check).
- additionally we are indexing period data by period blocks root now, so that later we can check with db if the node saw the period already.
- the node doesn't allow multiple submissions for the same blocks anymore. We used to store an array of submissions per blocks interval, now it is just one submission. This submission can be overwritten with the one with different blocks root.

## Breaking changes
- plasma_getPeriodByBlockHeight returns object not array
- storing a single Period submission per period now. New submission for the same starting height will overwrite, unless having the same blocksRoot

## How to test
Use https://github.com/leapdao/leap-sandbox/pull/80 version of sandbox

1. Start a root chain
2. Start a single validator node, submit couple of periods
3. Start a second node
4. Check period data on the second node (e.g. using tool https://github.com/leapdao/leap-guardian/pull/39). Should be identical to the first node.

# 2. Period proposals were not restored correctly on node restart #382

Main changes to solve this:
- check if the period  was processed already or not by querying database (see `bridgeState. getPeriodDataByBlocksRoot`) instead of checking `lastPeriodRoot` only (when replayed it will always has the very last period root)
- store "last seen" period root in database and use it to recreate `currentPeriod` on restart. It is needed because we are not replaying the chain and bridge events from scratch anymore.

## How to test
Use https://github.com/leapdao/leap-sandbox/pull/80 version of sandbox


----

Fixes #395, #382 
Requires https://github.com/leapdao/leap-core/pull/171, https://github.com/leapdao/leap-sandbox/pull/80
